// ngx-slimscroll/esm5/lib/slimscroll.directive.js
import { __assign, __decorate, __param, __read, __spread } from "tslib";
import { DOCUMENT } from '@angular/common';
import { Directive, EventEmitter, HostListener, Inject, Input, OnChanges, OnDestroy, OnInit, Optional, Output, Renderer2, SimpleChanges, ViewContainerRef } from '@angular/core';
import { fromEvent, merge, Subscription } from 'rxjs';
import { map, mergeMap, takeUntil } from 'rxjs/operators';
import { SlimScrollOptions, SLIMSCROLL_DEFAULTS } from './slimscroll-options.class';
import { SlimScrollState } from './slimscroll-state.class';
var ɵ0 = function (t) { return t; }, ɵ1 = function (t) { return t * t; }, ɵ2 = function (t) { return t * (2 - t); }, ɵ3 = function (t) { return (t < 0.5 ? 2 * t * t : -1 + (4 - 2 * t) * t); }, ɵ4 = function (t) { return t * t * t; }, ɵ5 = function (t) { return --t * t * t + 1; }, ɵ6 = function (t) { return (t < 0.5 ? 4 * t * t * t : (t - 1) * (2 * t - 2) * (2 * t - 2) + 1); }, ɵ7 = function (t) { return t * t * t * t; }, ɵ8 = function (t) { return 1 - --t * t * t * t; }, ɵ9 = function (t) { return (t < 0.5 ? 8 * t * t * t * t : 1 - 8 * --t * t * t * t); }, ɵ10 = function (t) { return t * t * t * t * t; }, ɵ11 = function (t) { return 1 + --t * t * t * t * t; }, ɵ12 = function (t) { return (t < 0.5 ? 16 * t * t * t * t * t : 1 + 16 * --t * t * t * t * t); };
export var easing = {
    linear: ɵ0,
    inQuad: ɵ1,
    outQuad: ɵ2,
    inOutQuad: ɵ3,
    inCubic: ɵ4,
    outCubic: ɵ5,
    inOutCubic: ɵ6,
    inQuart: ɵ7,
    outQuart: ɵ8,
    inOutQuart: ɵ9,
    inQuint: ɵ10,
    outQuint: ɵ11,
    inOutQuint: ɵ12
};
var SlimScrollDirective = /** @class */ (function () {
    function SlimScrollDirective(viewContainer, renderer, document, optionsDefaults) {
        this.viewContainer = viewContainer;
        this.renderer = renderer;
        this.document = document;
        this.optionsDefaults = optionsDefaults;
        this.enabled = true;
        this.scrollChanged = new EventEmitter();
        this.barVisibilityChange = new EventEmitter();
        this.el = this.viewContainer.element.nativeElement;
        this.body = this.document.querySelector('body');
        this.mutationThrottleTimeout = 50;
    }
    SlimScrollDirective.prototype.ngOnInit = function () {
        if (!this.interactionSubscriptions && this.enabled) {
            this.setup();
        }
    };
    SlimScrollDirective.prototype.ngOnChanges = function (changes) {
        if (changes.enabled) {
            if (this.enabled) {
                this.setup();
            }
            else {
                this.destroy();
            }
        }
    };
    SlimScrollDirective.prototype.ngOnDestroy = function () {
        this.destroy();
    };
    SlimScrollDirective.prototype.setup = function () {
        var _this = this;
        this.interactionSubscriptions = new Subscription();
        if (this.optionsDefaults) {
            this.options = new SlimScrollOptions(this.optionsDefaults).merge(this.options);
        }
        else {
            this.options = new SlimScrollOptions(this.options);
        }
        this.setStyle();
        this.wrapContainer();
        this.initGrid();
        this.initBar();
        this.getBarHeight();
        this.initWheel();
        this.initDrag();
        if (!this.options.alwaysVisible) {
            this.hideBarAndGrid();
        }
        if (MutationObserver) {
            if (this.mutationObserver) {
                this.mutationObserver.disconnect();
            }
            this.mutationObserver = new MutationObserver(function () {
                if (_this.mutationThrottleTimeout) {
                    clearTimeout(_this.mutationThrottleTimeout);
                    _this.mutationThrottleTimeout = setTimeout(_this.onMutation.bind(_this), 50);
                }
            });
            this.mutationObserver.observe(this.el, { subtree: true, childList: true });
        }
        if (this.scrollEvents && this.scrollEvents instanceof EventEmitter) {
            var scrollSubscription = this.scrollEvents.subscribe(function (event) { return _this.handleEvent(event); });
            this.interactionSubscriptions.add(scrollSubscription);
        }
    };
    SlimScrollDirective.prototype.handleEvent = function (e) {
        if (e.type === 'scrollToBottom') {
            var y = this.el.scrollHeight - this.el.clientHeight;
            this.scrollTo(y, e.duration, e.easing);
        }
        else if (e.type === 'scrollToTop') {
            var y = 0;
            this.scrollTo(y, e.duration, e.easing);
        }
        else if (e.type === 'scrollToPercent' && e.percent >= 0 && e.percent <= 100) {
            var y = Math.round(((this.el.scrollHeight - this.el.clientHeight) / 100) * e.percent);
            this.scrollTo(y, e.duration, e.easing);
        }
        else if (e.type === 'scrollTo') {
            var max = this.el.scrollHeight - this.el.clientHeight;
            var y = e.y <= max ? e.y : max;
            if (y >= 0) {
                this.scrollTo(y, e.duration, e.easing);
            }
        }
        else if (e.type === 'recalculate') {
            this.getBarHeight();
        }
    };
    SlimScrollDirective.prototype.setStyle = function () {
        var el = this.el;
        this.renderer.setStyle(el, 'overflow', 'hidden');
    };
    SlimScrollDirective.prototype.onMutation = function () {
        this.getBarHeight();
        this.setBarTop();
    };
    SlimScrollDirective.prototype.wrapContainer = function () {
        this.wrapper = this.renderer.createElement('div');
        var wrapper = this.wrapper;
        var el = this.el;
        this.renderer.addClass(wrapper, 'slimscroll-wrapper');
        this.renderer.setStyle(wrapper, 'position', 'relative');
        this.renderer.setStyle(wrapper, 'overflow', 'hidden');
        this.renderer.setStyle(wrapper, 'display', 'block');
        this.renderer.setStyle(wrapper, 'margin', getComputedStyle(el).margin);
        this.renderer.setStyle(wrapper, 'width', '100%');
        this.renderer.setStyle(wrapper, 'height', getComputedStyle(el).height);
        this.renderer.insertBefore(el.parentNode, wrapper, el);
        this.renderer.appendChild(wrapper, el);
    };
    SlimScrollDirective.prototype.initGrid = function () {
        this.grid = this.renderer.createElement('div');
        var grid = this.grid;
        this.renderer.addClass(grid, 'slimscroll-grid');
        this.renderer.setStyle(grid, 'position', 'absolute');
        this.renderer.setStyle(grid, 'top', '0');
        this.renderer.setStyle(grid, 'bottom', '0');
        this.renderer.setStyle(grid, this.options.position, '0');
        this.renderer.setStyle(grid, 'width', this.options.gridWidth + "px");
        this.renderer.setStyle(grid, 'background', this.options.gridBackground);
        this.renderer.setStyle(grid, 'opacity', this.options.gridOpacity);
        this.renderer.setStyle(grid, 'display', 'block');
        this.renderer.setStyle(grid, 'cursor', 'pointer');
        this.renderer.setStyle(grid, 'z-index', '99');
        this.renderer.setStyle(grid, 'border-radius', this.options.gridBorderRadius + "px");
        this.renderer.setStyle(grid, 'margin', this.options.gridMargin);
        this.renderer.appendChild(this.wrapper, grid);
    };
    SlimScrollDirective.prototype.initBar = function () {
        this.bar = this.renderer.createElement('div');
        var bar = this.bar;
        this.renderer.addClass(bar, 'slimscroll-bar');
        this.renderer.setStyle(bar, 'position', 'absolute');
        this.renderer.setStyle(bar, 'top', '0');
        this.renderer.setStyle(bar, this.options.position, '0');
        this.renderer.setStyle(bar, 'width', this.options.barWidth + "px");
        this.renderer.setStyle(bar, 'background', this.options.barBackground);
        this.renderer.setStyle(bar, 'opacity', this.options.barOpacity);
        this.renderer.setStyle(bar, 'display', 'block');
        this.renderer.setStyle(bar, 'cursor', 'pointer');
        this.renderer.setStyle(bar, 'z-index', '100');
        this.renderer.setStyle(bar, 'border-radius', this.options.barBorderRadius + "px");
        this.renderer.setStyle(bar, 'margin', this.options.barMargin);
        this.renderer.appendChild(this.wrapper, bar);
        this.barVisibilityChange.emit(true);
    };
    SlimScrollDirective.prototype.getBarHeight = function () {
        var elHeight = this.el.offsetHeight;
        var barHeight = Math.max((elHeight / this.el.scrollHeight) * elHeight, 30) + 'px';
        var display = parseInt(barHeight, 10) === elHeight ? 'none' : 'block';
        if (this.wrapper.offsetHeight !== elHeight) {
            this.renderer.setStyle(this.wrapper, 'height', elHeight + 'px');
        }
        this.renderer.setStyle(this.bar, 'height', barHeight);
        this.renderer.setStyle(this.bar, 'display', display);
        this.renderer.setStyle(this.grid, 'display', display);
        this.barVisibilityChange.emit(display !== 'none');
    };
    SlimScrollDirective.prototype.scrollTo = function (y, duration, easingFunc) {
        var _this = this;
        var start = Date.now();
        var from = this.el.scrollTop;
        var paddingTop = parseInt(this.el.style.paddingTop, 10) || 0;
        var paddingBottom = parseInt(this.el.style.paddingBottom, 10) || 0;
        var scroll = function () {
            var currentTime = Date.now();
            var time = Math.min(1, (currentTime - start) / duration);
            var easedTime = easing[easingFunc](time);
            if (paddingTop > 0 || paddingBottom > 0) {
                var fromY = null;
                if (paddingTop > 0) {
                    fromY = -paddingTop;
                    fromY = -(easedTime * (y - fromY) + fromY);
                    _this.renderer.setStyle(_this.el, 'padding-top', fromY + "px");
                }
                if (paddingBottom > 0) {
                    fromY = paddingBottom;
                    fromY = easedTime * (y - fromY) + fromY;
                    _this.renderer.setStyle(_this.el, 'padding-bottom', fromY + "px");
                }
            }
            else {
                _this.el.scrollTop = easedTime * (y - from) + from;
            }
            _this.setBarTop();
            _this.saveCurrent();
            _this.updateScrollState();
            if (time < 1) {
                requestAnimationFrame(scroll);
            }
        };
        if (!duration || !easingFunc) {
            this.el.scrollTop = y;
            this.setBarTop();
            this.saveCurrent();
            this.updateScrollState();
        }
        else {
            requestAnimationFrame(scroll);
        }
    };
    SlimScrollDirective.prototype.scrollContent = function (y, isWheel, isJump) {
        var _this = this;
        var delta = y;
        var maxTop = this.el.offsetHeight - this.bar.offsetHeight;
        var hiddenContent = this.el.scrollHeight - this.el.offsetHeight;
        var percentScroll;
        var over = null;
        if (isWheel) {
            delta = parseInt(getComputedStyle(this.bar).top, 10) + ((y * 20) / 100) * this.bar.offsetHeight;
            if (delta < 0 || delta > maxTop) {
                over = delta > maxTop ? delta - maxTop : delta;
            }
            delta = Math.min(Math.max(delta, 0), maxTop);
            delta = y > 0 ? Math.ceil(delta) : Math.floor(delta);
            this.renderer.setStyle(this.bar, 'top', delta + 'px');
        }
        percentScroll = parseInt(getComputedStyle(this.bar).top, 10) / (this.el.offsetHeight - this.bar.offsetHeight);
        delta = percentScroll * hiddenContent;
        this.el.scrollTop = delta;
        this.showBarAndGrid();
        if (!this.options.alwaysVisible) {
            if (this.visibleTimeout) {
                clearTimeout(this.visibleTimeout);
            }
            this.visibleTimeout = setTimeout(function () {
                _this.hideBarAndGrid();
            }, this.options.visibleTimeout);
        }
        this.updateScrollState();
        this.saveCurrent();
        return over;
    };
    SlimScrollDirective.prototype.updateScrollState = function () {
        var isScrollAtStart = this.el.scrollTop === 0;
        var isScrollAtEnd = this.el.scrollTop === this.el.scrollHeight - this.el.offsetHeight;
        var scrollPosition = Math.ceil(this.el.scrollTop);
        var scrollState = new SlimScrollState({ scrollPosition: scrollPosition, isScrollAtStart: isScrollAtStart, isScrollAtEnd: isScrollAtEnd });
        this.scrollChanged.emit(scrollState);
    };
    SlimScrollDirective.prototype.initWheel = function () {
        var _this = this;
        var dommousescroll = fromEvent(this.el, 'DOMMouseScroll');
        var mousewheel = fromEvent(this.el, 'mousewheel');
        var wheelSubscription = merge.apply(void 0, __spread([dommousescroll, mousewheel])).subscribe(function (e) {
            var delta = 0;
            if (e.wheelDelta) {
                delta = -e.wheelDelta / 120;
            }
            if (e.detail) {
                delta = e.detail / 3;
            }
            var over = _this.scrollContent(delta, true, false);
            if (e.preventDefault && (_this.options.alwaysPreventDefaultScroll || over === null)) {
                e.preventDefault();
            }
        });
        this.interactionSubscriptions.add(wheelSubscription);
    };
    SlimScrollDirective.prototype.initDrag = function () {
        var _this = this;
        var bar = this.bar;
        var mousemove = fromEvent(this.document.documentElement, 'mousemove');
        var touchmove = fromEvent(this.document.documentElement, 'touchmove');
        var mousedown = fromEvent(bar, 'mousedown');
        var touchstart = fromEvent(this.el, 'touchstart');
        var mouseup = fromEvent(this.document.documentElement, 'mouseup');
        var touchend = fromEvent(this.document.documentElement, 'touchend');
        var mousedrag = mousedown.pipe(mergeMap(function (e) {
            _this.pageY = e.pageY;
            _this.top = parseFloat(getComputedStyle(bar).top);
            return mousemove.pipe(map(function (emove) {
                emove.preventDefault();
                return _this.top + emove.pageY - _this.pageY;
            }), takeUntil(mouseup));
        }));
        var touchdrag = touchstart.pipe(mergeMap(function (e) {
            _this.pageY = e.targetTouches[0].pageY;
            _this.top = -parseFloat(getComputedStyle(bar).top);
            return touchmove.pipe(map(function (tmove) {
                return -(_this.top + tmove.targetTouches[0].pageY - _this.pageY);
            }), takeUntil(touchend));
        }));
        var dragSubscription = merge.apply(void 0, __spread([mousedrag, touchdrag])).subscribe(function (top) {
            _this.body.addEventListener('selectstart', _this.preventDefaultEvent, false);
            _this.renderer.setStyle(_this.body, 'touch-action', 'pan-y');
            _this.renderer.setStyle(_this.body, 'user-select', 'none');
            _this.renderer.setStyle(_this.bar, 'top', top + "px");
            var over = _this.scrollContent(0, true, false);
            var maxTop = _this.el.offsetHeight - _this.bar.offsetHeight;
            if (over && over < 0 && -over <= maxTop) {
                _this.renderer.setStyle(_this.el, 'paddingTop', -over + 'px');
            }
            else if (over && over > 0 && over <= maxTop) {
                _this.renderer.setStyle(_this.el, 'paddingBottom', over + 'px');
            }
        });
        var dragEndSubscription = merge.apply(void 0, __spread([mouseup, touchend])).subscribe(function () {
            _this.body.removeEventListener('selectstart', _this.preventDefaultEvent, false);
            var paddingTop = parseInt(_this.el.style.paddingTop, 10);
            var paddingBottom = parseInt(_this.el.style.paddingBottom, 10);
            _this.renderer.setStyle(_this.body, 'touch-action', 'unset');
            _this.renderer.setStyle(_this.body, 'user-select', 'default');
            if (paddingTop > 0) {
                _this.scrollTo(0, 300, 'linear');
            }
            else if (paddingBottom > 0) {
                _this.scrollTo(0, 300, 'linear');
            }
        });
        this.interactionSubscriptions.add(dragSubscription);
        this.interactionSubscriptions.add(dragEndSubscription);
    };
    SlimScrollDirective.prototype.setBarTop = function () {
        var barHeight = Math.max((this.el.offsetHeight / this.el.scrollHeight) * this.el.offsetHeight, 30);
        var maxScrollTop = this.el.scrollHeight - this.el.clientHeight;
        var paddingBottom = parseInt(this.el.style.paddingBottom, 10) || 0;
        var percentScroll = this.el.scrollTop / maxScrollTop;
        if (paddingBottom === 0) {
            var delta = Math.round((this.el.clientHeight - barHeight) * percentScroll);
            if (delta > 0) {
                this.renderer.setStyle(this.bar, 'top', delta + "px");
            }
        }
    };
    SlimScrollDirective.prototype.saveCurrent = function () {
        var max = this.el.scrollHeight - this.el.clientHeight;
        var percent = this.el.scrollTop / max;
        this.current = { max: max, percent: percent };
    };
    SlimScrollDirective.prototype.showBarAndGrid = function () {
        this.renderer.setStyle(this.grid, 'background', this.options.gridBackground);
        this.renderer.setStyle(this.bar, 'background', this.options.barBackground);
    };
    SlimScrollDirective.prototype.hideBarAndGrid = function () {
        this.renderer.setStyle(this.grid, 'background', 'transparent');
        this.renderer.setStyle(this.bar, 'background', 'transparent');
    };
    SlimScrollDirective.prototype.preventDefaultEvent = function (e) {
        e.preventDefault();
        e.stopPropagation();
    };
    SlimScrollDirective.prototype.destroy = function () {
        if (this.mutationObserver) {
            this.mutationObserver.disconnect();
            this.mutationObserver = null;
        }
        if (this.el.parentElement.classList.contains('slimscroll-wrapper')) {
            var wrapper = this.el.parentElement;
            var bar = wrapper.querySelector('.slimscroll-bar');
            wrapper.removeChild(bar);
            var grid = wrapper.querySelector('.slimscroll-grid');
            wrapper.removeChild(grid);
            this.unwrap(wrapper);
        }
        if (this.interactionSubscriptions) {
            this.interactionSubscriptions.unsubscribe();
        }
    };
    SlimScrollDirective.prototype.unwrap = function (wrapper) {
        var docFrag = document.createDocumentFragment();
        while (wrapper.firstChild) {
            var child = wrapper.removeChild(wrapper.firstChild);
            docFrag.appendChild(child);
        }
        wrapper.parentNode.replaceChild(docFrag, wrapper);
    };
    SlimScrollDirective.prototype.onResize = function () {
        var percent = __assign({}, this.current).percent;
        this.destroy();
        this.setup();
        this.scrollTo(Math.round((this.el.scrollHeight - this.el.clientHeight) * percent), null, null);
    };
    SlimScrollDirective.ctorParameters = function () { return [
        { type: ViewContainerRef, decorators: [{ type: Inject, args: [ViewContainerRef,] }] },
        { type: Renderer2, decorators: [{ type: Inject, args: [Renderer2,] }] },
        { type: undefined, decorators: [{ type: Inject, args: [DOCUMENT,] }] },
        { type: undefined, decorators: [{ type: Inject, args: [SLIMSCROLL_DEFAULTS,] }, { type: Optional }] }
    ]; };
    __decorate([
        Input()
    ], SlimScrollDirective.prototype, "enabled", void 0);
    __decorate([
        Input()
    ], SlimScrollDirective.prototype, "options", void 0);
    __decorate([
        Input()
    ], SlimScrollDirective.prototype, "scrollEvents", void 0);
    __decorate([
        Output()
    ], SlimScrollDirective.prototype, "scrollChanged", void 0);
    __decorate([
        Output()
    ], SlimScrollDirective.prototype, "barVisibilityChange", void 0);
    __decorate([
        HostListener('window:resize', [])
    ], SlimScrollDirective.prototype, "onResize", null);
    SlimScrollDirective = __decorate([
        Directive({
            selector: '[slimScroll]',
            exportAs: 'slimScroll'
        }),
        __param(0, Inject(ViewContainerRef)),
        __param(1, Inject(Renderer2)),
        __param(2, Inject(DOCUMENT)),
        __param(3, Inject(SLIMSCROLL_DEFAULTS)), __param(3, Optional())
    ], SlimScrollDirective);
    return SlimScrollDirective;
}());
export { SlimScrollDirective };
export { ɵ0, ɵ1, ɵ2, ɵ3, ɵ4, ɵ5, ɵ6, ɵ7, ɵ8, ɵ9, ɵ10, ɵ11, ɵ12 };
