// angular-ng-stepper/esm5/lib/ng-stepper.component.js
import { __decorate, __extends } from "tslib";
import { Component, HostListener } from '@angular/core';
import { CdkStepper, STEPPER_GLOBAL_OPTIONS } from '@angular/cdk/stepper';
import { animate, state, style, transition, trigger } from '@angular/animations';
var ɵ0 = { displayDefaultIndicatorType: false };
var NgStepperComponent = /** @class */ (function (_super) {
    __extends(NgStepperComponent, _super);
    function NgStepperComponent() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    NgStepperComponent_1 = NgStepperComponent;
    NgStepperComponent.prototype.ngOnInit = function () {
    };
    NgStepperComponent.prototype.handleKeyboardEvent = function (event) {
        if (event.key === 'ArrowLeft') {
            this.previous();
        }
        else if (event.key === 'ArrowRight') {
            this.next();
        }
    };
    var NgStepperComponent_1;
    __decorate([
        HostListener('document:keydown', ['$event'])
    ], NgStepperComponent.prototype, "handleKeyboardEvent", null);
    NgStepperComponent = NgStepperComponent_1 = __decorate([
        Component({
            selector: 'ng-stepper',
            template: "<div class=\"stepper\">\n  <div class=\"stepper-header\">\n    <ol class=\"steps\">\n      <ng-container *ngFor=\"let step of steps; let i = index;\">\n        <li class=\"step\" (click)=\"selectedIndex = i\"\n            [ngClass]=\"{\n                active: selectedIndex === i,\n                completed: step.completed && i < selectedIndex,\n                done: step.state === 'done' && i === selectedIndex\n                }\">\n          <ng-container [ngTemplateOutlet]=\"step.stepLabel.template\">\n            <ng-content select=\".step-bullet\"></ng-content>\n            <ng-content select=\".step-title\"></ng-content>\n          </ng-container>\n        </li>\n      </ng-container>\n    </ol>\n  </div>\n  <div class=\"stepper-body\">\n    <div class=\"content\"\n         *ngFor=\"let step of _steps; let i = index\"\n         [class.active]=\"selectedIndex === i\"\n         [@stepTransition]=\"_getAnimationDirection(i)\">\n      <ng-container [ngTemplateOutlet]=\"step.content\"></ng-container>\n    </div>\n  </div>\n</div>\n\n",
            providers: [{ provide: CdkStepper, useExisting: NgStepperComponent_1 }, {
                    provide: STEPPER_GLOBAL_OPTIONS, useValue: ɵ0
                }],
            animations: [trigger('stepTransition', [
                    state('previous', style({ transform: 'translateX(-100%)', opacity: 0 })),
                    state('current', style({ transform: 'translateX(0)', opacity: 1 })),
                    state('next', style({ transform: 'translateX(100%)', opacity: 0 })),
                    transition(':enter', animate(0)),
                    transition('previous => current', animate('500ms cubic-bezier(0.35, 0, 0.25, 1)')),
                    transition('next => current', animate('500ms cubic-bezier(0.35, 0, 0.25, 1)')),
                ])],
            styles: [":host{display:block;overflow:hidden}.stepper-header .steps{width:100%;list-style:none;display:inline-flex;align-items:center;margin:0 auto}.stepper-header .steps .step{text-align:center;font-size:1em}.stepper-header .steps ::ng-deep .step .step-bullet{position:relative;display:inline-flex;justify-content:center;align-items:center;height:40px;width:40px;background-color:#b3babe;border-radius:50%;color:#fff;font-size:1.3rem;text-align:center;text-decoration:none;cursor:pointer}.stepper-header .steps ::ng-deep .step .step-title{position:absolute;right:0;bottom:-5px;left:0;margin:auto 0;height:0}.stepper-header .steps ::ng-deep .step.active .step-bullet{background-color:#306bff}.stepper-header .steps ::ng-deep .step.completed .step-bullet{background-color:#32e67f}.stepper-header .steps ::ng-deep .step.done .step-bullet{background-color:#32e67f}.stepper-body{margin:25px auto;padding:1rem}.stepper-body .content:not(.active){height:0;overflow:hidden}.step{position:relative;width:33.33%}.step:after{content:\"\";position:absolute;width:100%;height:3px;background-color:#b3babe;top:50%;left:-50%;z-index:-1}.step:first-child:after{content:none}.step.completed+.step:after{background-color:#32e67f}"]
        })
    ], NgStepperComponent);
    return NgStepperComponent;
}(CdkStepper));
export { NgStepperComponent };
export { ɵ0 };
