// angular-ng-stepper/__ivy_ngcc__/fesm2015/angular-ng-stepper.js
import { __decorate } from 'tslib';
import { HostListener, Component, NgModule } from '@angular/core';
import { CdkStepper, STEPPER_GLOBAL_OPTIONS, CdkStepperModule } from '@angular/cdk/stepper';
import { trigger, state, style, transition, animate } from '@angular/animations';
import { CommonModule } from '@angular/common';

import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from '@angular/common';

const _c0 = function (a0, a1, a2) { return { active: a0, completed: a1, done: a2 }; };
function NgStepperComponent_ng_container_3_Template(rf, ctx) { if (rf & 1) {
    const _r5 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementContainerStart(0);
    ɵngcc0.ɵɵelementStart(1, "li", 6);
    ɵngcc0.ɵɵlistener("click", function NgStepperComponent_ng_container_3_Template_li_click_1_listener() { ɵngcc0.ɵɵrestoreView(_r5); const i_r3 = ctx.index; const ctx_r4 = ɵngcc0.ɵɵnextContext(); return ctx_r4.selectedIndex = i_r3; });
    ɵngcc0.ɵɵelementContainerStart(2, 7);
    ɵngcc0.ɵɵprojection(3);
    ɵngcc0.ɵɵprojection(4, 1);
    ɵngcc0.ɵɵelementContainerEnd();
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementContainerEnd();
} if (rf & 2) {
    const step_r2 = ctx.$implicit;
    const i_r3 = ctx.index;
    const ctx_r0 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngClass", ɵngcc0.ɵɵpureFunction3(2, _c0, ctx_r0.selectedIndex === i_r3, step_r2.completed && i_r3 < ctx_r0.selectedIndex, step_r2.state === "done" && i_r3 === ctx_r0.selectedIndex));
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngTemplateOutlet", step_r2.stepLabel.template);
} }
function NgStepperComponent_div_5_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "div", 8);
    ɵngcc0.ɵɵelementContainer(1, 7);
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const step_r6 = ctx.$implicit;
    const i_r7 = ctx.index;
    const ctx_r1 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵclassProp("active", ctx_r1.selectedIndex === i_r7);
    ɵngcc0.ɵɵproperty("@stepTransition", ctx_r1._getAnimationDirection(i_r7));
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("ngTemplateOutlet", step_r6.content);
} }
const _c1 = [[["", 8, "step-bullet"]], [["", 8, "step-title"]]];
const _c2 = [".step-bullet", ".step-title"];
var NgStepperComponent_1;
const ɵ0 = { displayDefaultIndicatorType: false };
let NgStepperComponent = NgStepperComponent_1 = class NgStepperComponent extends CdkStepper {
    ngOnInit() {
    }
    handleKeyboardEvent(event) {
        if (event.key === 'ArrowLeft') {
            this.previous();
        }
        else if (event.key === 'ArrowRight') {
            this.next();
        }
    }
};
NgStepperComponent.ɵfac = function NgStepperComponent_Factory(t) { return ɵNgStepperComponent_BaseFactory(t || NgStepperComponent); };
NgStepperComponent.ɵcmp = ɵngcc0.ɵɵdefineComponent({ type: NgStepperComponent, selectors: [["ng-stepper"]], hostBindings: function NgStepperComponent_HostBindings(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵlistener("keydown", function NgStepperComponent_keydown_HostBindingHandler($event) { return ctx.handleKeyboardEvent($event); }, false, ɵngcc0.ɵɵresolveDocument);
    } }, features: [ɵngcc0.ɵɵProvidersFeature([{ provide: CdkStepper, useExisting: NgStepperComponent_1 }, {
                provide: STEPPER_GLOBAL_OPTIONS, useValue: ɵ0
            }]), ɵngcc0.ɵɵInheritDefinitionFeature], ngContentSelectors: _c2, decls: 6, vars: 2, consts: [[1, "stepper"], [1, "stepper-header"], [1, "steps"], [4, "ngFor", "ngForOf"], [1, "stepper-body"], ["class", "content", 3, "active", 4, "ngFor", "ngForOf"], [1, "step", 3, "ngClass", "click"], [3, "ngTemplateOutlet"], [1, "content"]], template: function NgStepperComponent_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵprojectionDef(_c1);
        ɵngcc0.ɵɵelementStart(0, "div", 0);
        ɵngcc0.ɵɵelementStart(1, "div", 1);
        ɵngcc0.ɵɵelementStart(2, "ol", 2);
        ɵngcc0.ɵɵtemplate(3, NgStepperComponent_ng_container_3_Template, 5, 6, "ng-container", 3);
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(4, "div", 4);
        ɵngcc0.ɵɵtemplate(5, NgStepperComponent_div_5_Template, 2, 4, "div", 5);
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementEnd();
    } if (rf & 2) {
        ɵngcc0.ɵɵadvance(3);
        ɵngcc0.ɵɵproperty("ngForOf", ctx.steps);
        ɵngcc0.ɵɵadvance(2);
        ɵngcc0.ɵɵproperty("ngForOf", ctx._steps);
    } }, directives: [ɵngcc1.NgForOf, ɵngcc1.NgClass, ɵngcc1.NgTemplateOutlet], styles: ["[_nghost-%COMP%]{display:block;overflow:hidden}.stepper-header[_ngcontent-%COMP%]   .steps[_ngcontent-%COMP%]{width:100%;list-style:none;display:inline-flex;align-items:center;margin:0 auto}.stepper-header[_ngcontent-%COMP%]   .steps[_ngcontent-%COMP%]   .step[_ngcontent-%COMP%]{text-align:center;font-size:1em}.stepper-header[_ngcontent-%COMP%]   .steps[_ngcontent-%COMP%]     .step .step-bullet{position:relative;display:inline-flex;justify-content:center;align-items:center;height:40px;width:40px;background-color:#b3babe;border-radius:50%;color:#fff;font-size:1.3rem;text-align:center;text-decoration:none;cursor:pointer}.stepper-header[_ngcontent-%COMP%]   .steps[_ngcontent-%COMP%]     .step .step-title{position:absolute;right:0;bottom:-5px;left:0;margin:auto 0;height:0}.stepper-header[_ngcontent-%COMP%]   .steps[_ngcontent-%COMP%]     .step.active .step-bullet{background-color:#306bff}.stepper-header[_ngcontent-%COMP%]   .steps[_ngcontent-%COMP%]     .step.completed .step-bullet{background-color:#32e67f}.stepper-header[_ngcontent-%COMP%]   .steps[_ngcontent-%COMP%]     .step.done .step-bullet{background-color:#32e67f}.stepper-body[_ngcontent-%COMP%]{margin:25px auto;padding:1rem}.stepper-body[_ngcontent-%COMP%]   .content[_ngcontent-%COMP%]:not(.active){height:0;overflow:hidden}.step[_ngcontent-%COMP%]{position:relative;width:33.33%}.step[_ngcontent-%COMP%]:after{content:\"\";position:absolute;width:100%;height:3px;background-color:#b3babe;top:50%;left:-50%;z-index:-1}.step[_ngcontent-%COMP%]:first-child:after{content:none}.step.completed[_ngcontent-%COMP%] + .step[_ngcontent-%COMP%]:after{background-color:#32e67f}"], data: { animation: [trigger('stepTransition', [
                state('previous', style({ transform: 'translateX(-100%)', opacity: 0 })),
                state('current', style({ transform: 'translateX(0)', opacity: 1 })),
                state('next', style({ transform: 'translateX(100%)', opacity: 0 })),
                transition(':enter', animate(0)),
                transition('previous => current', animate('500ms cubic-bezier(0.35, 0, 0.25, 1)')),
                transition('next => current', animate('500ms cubic-bezier(0.35, 0, 0.25, 1)')),
            ])] } });
__decorate([
    HostListener('document:keydown', ['$event'])
], NgStepperComponent.prototype, "handleKeyboardEvent", null);

let NgStepperModule = class NgStepperModule {
};
NgStepperModule.ɵfac = function NgStepperModule_Factory(t) { return new (t || NgStepperModule)(); };
NgStepperModule.ɵmod = ɵngcc0.ɵɵdefineNgModule({ type: NgStepperModule });
NgStepperModule.ɵinj = ɵngcc0.ɵɵdefineInjector({ imports: [[
            CommonModule,
            CdkStepperModule
        ]] });
const ɵNgStepperComponent_BaseFactory = /*@__PURE__*/ ɵngcc0.ɵɵgetInheritedFactory(NgStepperComponent);
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(NgStepperComponent, [{
        type: Component,
        args: [{
                selector: 'ng-stepper',
                template: "<div class=\"stepper\">\n  <div class=\"stepper-header\">\n    <ol class=\"steps\">\n      <ng-container *ngFor=\"let step of steps; let i = index;\">\n        <li class=\"step\" (click)=\"selectedIndex = i\"\n            [ngClass]=\"{\n                active: selectedIndex === i,\n                completed: step.completed && i < selectedIndex,\n                done: step.state === 'done' && i === selectedIndex\n                }\">\n          <ng-container [ngTemplateOutlet]=\"step.stepLabel.template\">\n            <ng-content select=\".step-bullet\"></ng-content>\n            <ng-content select=\".step-title\"></ng-content>\n          </ng-container>\n        </li>\n      </ng-container>\n    </ol>\n  </div>\n  <div class=\"stepper-body\">\n    <div class=\"content\"\n         *ngFor=\"let step of _steps; let i = index\"\n         [class.active]=\"selectedIndex === i\"\n         [@stepTransition]=\"_getAnimationDirection(i)\">\n      <ng-container [ngTemplateOutlet]=\"step.content\"></ng-container>\n    </div>\n  </div>\n</div>\n\n",
                providers: [{ provide: CdkStepper, useExisting: NgStepperComponent_1 }, {
                        provide: STEPPER_GLOBAL_OPTIONS, useValue: ɵ0
                    }],
                animations: [trigger('stepTransition', [
                        state('previous', style({ transform: 'translateX(-100%)', opacity: 0 })),
                        state('current', style({ transform: 'translateX(0)', opacity: 1 })),
                        state('next', style({ transform: 'translateX(100%)', opacity: 0 })),
                        transition(':enter', animate(0)),
                        transition('previous => current', animate('500ms cubic-bezier(0.35, 0, 0.25, 1)')),
                        transition('next => current', animate('500ms cubic-bezier(0.35, 0, 0.25, 1)')),
                    ])],
                styles: [":host{display:block;overflow:hidden}.stepper-header .steps{width:100%;list-style:none;display:inline-flex;align-items:center;margin:0 auto}.stepper-header .steps .step{text-align:center;font-size:1em}.stepper-header .steps ::ng-deep .step .step-bullet{position:relative;display:inline-flex;justify-content:center;align-items:center;height:40px;width:40px;background-color:#b3babe;border-radius:50%;color:#fff;font-size:1.3rem;text-align:center;text-decoration:none;cursor:pointer}.stepper-header .steps ::ng-deep .step .step-title{position:absolute;right:0;bottom:-5px;left:0;margin:auto 0;height:0}.stepper-header .steps ::ng-deep .step.active .step-bullet{background-color:#306bff}.stepper-header .steps ::ng-deep .step.completed .step-bullet{background-color:#32e67f}.stepper-header .steps ::ng-deep .step.done .step-bullet{background-color:#32e67f}.stepper-body{margin:25px auto;padding:1rem}.stepper-body .content:not(.active){height:0;overflow:hidden}.step{position:relative;width:33.33%}.step:after{content:\"\";position:absolute;width:100%;height:3px;background-color:#b3babe;top:50%;left:-50%;z-index:-1}.step:first-child:after{content:none}.step.completed+.step:after{background-color:#32e67f}"]
            }]
    }], null, { handleKeyboardEvent: [{
            type: HostListener,
            args: ['document:keydown', ['$event']]
        }] }); })();
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && ɵngcc0.ɵɵsetNgModuleScope(NgStepperModule, { declarations: function () { return [NgStepperComponent]; }, imports: function () { return [CommonModule,
        CdkStepperModule]; }, exports: function () { return [NgStepperComponent]; } }); })();
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(NgStepperModule, [{
        type: NgModule,
        args: [{
                declarations: [NgStepperComponent],
                imports: [
                    CommonModule,
                    CdkStepperModule
                ],
                exports: [NgStepperComponent]
            }]
    }], null, null); })();

/*
 * Public API Surface of ng-stepper
 */

/**
 * Generated bundle index. Do not edit.
 */

export { NgStepperComponent, NgStepperModule, ɵ0 };
